lib/api/api_auth.go

```diff
@@ -210,3 +315,91 @@ func iso88591ToUTF8(s []byte) []byte {
 	}
 	return []byte(string(runes))
 }
+
+type webauthnService struct {
+	webauthnState webauthnLib.SessionData
+	cfg config.Wrapper
+	cookieName string
+	evLogger events.Logger
+}
+
+func newWebauthnService(cfg config.Wrapper, cookieName string, evLogger events.Logger) webauthnService {
+	return webauthnService{
+		cfg: cfg,
+		cookieName: cookieName,
+		evLogger: evLogger,
+	}
+}
+
+func (s *webauthnService) startWebauthnAuthentication(w http.ResponseWriter, r *http.Request) {
+	webauthn, err := config.NewWebauthnHandle(s.cfg)
+	if err != nil {
+		l.Warnln("Failed to initialize WebAuthn handle", err)
+		internalServerError(w)
+		return
+	}
+
+	options, sessionData, err := webauthn.BeginLogin(s.cfg.GUI())
+	if err != nil {
+		badRequest, ok := err.(*webauthnProtocol.Error)
+		if ok && badRequest.Type == "invalid_request" && badRequest.Details == "Found no credentials for user" {
+			sendJSON(w, make(map[string]string))
+		} else {
+			l.Warnln("Failed to initialize WebAuthn login", err)
+		}
+		return
+	}
+
+	s.webauthnState = *sessionData
+
+	sendJSON(w, options)
+}
+
+func (s *webauthnService) finishWebauthnAuthentication(w http.ResponseWriter, r *http.Request) {
+	webauthn, err := config.NewWebauthnHandle(s.cfg)
+	if err != nil {
+		l.Warnln("Failed to initialize WebAuthn handle", err)
+		internalServerError(w)
+		return
+	}
+
+	state := s.webauthnState
+	s.webauthnState = webauthnLib.SessionData{} // Allow only one attempt per challenge
+
+	parsedResponse, err := webauthnProtocol.ParseCredentialRequestResponse(r)
+	if err != nil {
+		l.Debugln("Failed to parse WebAuthn authentication response", err)
+		badRequest(w)
+		return
+	}
+
+	guiCfg := s.cfg.GUI()
+	updatedCred, err := webauthn.ValidateLogin(guiCfg, state, parsedResponse)
+	if err != nil {
+		l.Infoln("WebAuthn authentication failed", err)
+		forbidden(w)
+		return
+	}
+
+	authenticatedCredId := base64.URLEncoding.EncodeToString(updatedCred.ID)
+	authenticatedCredName := authenticatedCredId
+	var signCountBefore uint32 = 0
+	waiter, err := s.cfg.Modify(func(cfg *config.Configuration) {
+		for i, cred := range cfg.GUI.WebauthnCredentials {
+			if cred.ID == authenticatedCredId {
+				signCountBefore = cfg.GUI.WebauthnCredentials[i].SignCount
+				authenticatedCredName = cfg.GUI.WebauthnCredentials[i].Nickname
+				cfg.GUI.WebauthnCredentials[i].SignCount = updatedCred.Authenticator.SignCount
+				break
+			}
+		}
+	})
+	s.cfg.Finish(w, waiter)
+
+	if updatedCred.Authenticator.CloneWarning && signCountBefore != 0 {
+		l.Warnln(fmt.Sprintf("Invalid WebAuthn signature count for credential \"%s\": expected > %d, was: %d. The credential may have been cloned.", authenticatedCredName, signCountBefore, parsedResponse.Response.AuthenticatorData.Counter))
+	}
```

Comment:
Can you explain what this counter mechanism does?

Also the messages mentions a comparison to the auth response thing, but the actual condition is just `> 0`.

And the counter isn't config, right? that should go into the db not the config.

Reply:
It's an optional feature that WebAuthn authenticators can implement, in which case every authentication signature by the credential will include a monotonically increasing counter value. So if a new authentication signature has a counter value less than or equal to the last seen counter value, then that means there likely exists two copies of the credential private key. But the feature is optional, and authenticators that don't implement it instead leave the counter constant at zero, so the check is only made if either the new or the last seen counter value is nonzero. You can read more about it [here](https://w3c.github.io/webauthn/#signature-counter). But the library [encapsulates all that in the `CloneWarning` output](https://github.com/duo-labs/webauthn/blob/092d5e0bbb16390d0cb46851c433218ac4a45d7a/webauthn/authenticator.go#L46-L47), so the zero check here is unrelated.

I think the `!= 0` check here is just because as far as I understand, the Go convention for searching in a list is to initialize to a sentinel value and use a for loop (unlike, say, Rust where you'd `.find(...).map(...) -> Option<u32>`). I'd love to be wrong about that and learn there's a better way, though. :slightly_smiling_face:

> And the counter isn't config, right? that should go into the db not the config.

Hmyeah, I guess so? In that case the whole credential entries should perhaps go in the database instead, as it's not really feasible to type them in manually either. How do you reckon that would work with configuring them in the settings view? Where should I start looking at how to work with the database?